pkg/eestream/decode.go

```diff
@@ -4,38 +4,96 @@
 
 	"storj.io/storj/internal/pkg/readcloser"
 	"storj.io/storj/pkg/ranger"
 )
 
 type decodedReader struct {
+	ctx    context.Context
+	cancel context.CancelFunc
 	rs     map[int]io.ReadCloser
 	es     ErasureScheme
-	inbufs map[int][]byte
 	outbuf []byte
 	err    error
+	chans  map[int]chan block
+	cb     int64 // current block number
+	eb     int64 // expected number of blocks
 }
 
-type readerError struct {
-	i   int // reader index in the map
-	err error
+type block struct {
+	i    int    // reader index in the map
+	num  int64  // block number
+	data []byte // block data
+	err  error  // error reading the block
 }
 
 // DecodeReaders takes a map of readers and an ErasureScheme returning a
 // combined Reader. The map, 'rs', must be a mapping of erasure piece numbers
-// to erasure piece streams.
-func DecodeReaders(rs map[int]io.ReadCloser, es ErasureScheme) io.ReadCloser {
+// to erasure piece streams. expectedSize is the number of bytes expected to
+// be returned by the Reader. maxBufferMemory is the maximum memory (in bytes)
+// to be allocated for read buffers. If set to 0, the minimum possible memory
+// will be used.
+func DecodeReaders(ctx context.Context, rs map[int]io.ReadCloser,
+	es ErasureScheme, expectedSize int64, maxBufferMemory int) io.ReadCloser {
+	if expectedSize < 0 {
+		return readcloser.FatalReadCloser(Error.New("negative expected size"))
+	}
+	if expectedSize%int64(es.DecodedBlockSize()) != 0 {
+		return readcloser.FatalReadCloser(
+			Error.New("expected size not a factor decoded block size"))
+	}
+	if maxBufferMemory < 0 {
+		return readcloser.FatalReadCloser(
+			Error.New("negative max buffer memory"))
+	}
+	chanSize := maxBufferMemory / (len(rs) * es.EncodedBlockSize())
+	if chanSize < 1 {
+		chanSize = 1
+	}
+	context, cancel := context.WithCancel(ctx)
 	dr := &decodedReader{
+		ctx:    context,
+		cancel: cancel,
 		rs:     rs,
 		es:     es,
-		inbufs: make(map[int][]byte, len(rs)),
 		outbuf: make([]byte, 0, es.DecodedBlockSize()),
+		chans:  make(map[int]chan block, len(rs)),
+		eb:     expectedSize / int64(es.DecodedBlockSize()),
 	}
+	// Kick off a goroutine for each reader. Each reads a block from the
+	// reader and adds it to a buffered channel. If an error is read
+	// (including EOF), a block with the error is added to the channel,
+	// the channel is closed and the goroutine exits.
 	for i := range rs {
-		dr.inbufs[i] = make([]byte, es.EncodedBlockSize())
+		dr.chans[i] = make(chan block, chanSize)
+		go func(i int, ch chan block) {
+			// close the channel when the goroutine exits
+			defer close(ch)
+			for blockNum := int64(0); ; blockNum++ {
+				// read the next block
+				data := make([]byte, es.EncodedBlockSize())
+				_, err := io.ReadFull(dr.rs[i], data)
+				// add it to the channel
+				select {
+				case ch <- block{i, blockNum, data, err}:
+					if err != nil {
+						// exit the goroutine (will close the channel)
+						return
+					}
+				case <-ctx.Done():
```

Comment:
so this is fun, but you're actually calling ctx.Done() every loop through the select loop. it's probably fine since the implementation of a context that context.WithCancel returns is safe for this, but just in general it's slightly more robust to call ctx.Done() once, then use the saved off value from that return in a loop.

no change required, just fyi

Reply:
Well, I just followed the official documentation of using it :)

Reply:
:+1: